File: backoffice/src/shared/infrastructure/eventBus/RabbitmqEventBusConfigurer.py
"""
 *
 * Libraries 
 *
"""

from injector                         import inject
from ..events.DomainEventsInformation import DomainEventsInformation
from amqpstorm                        import Channel
from .RabbitmqEventBusConnector       import RabbitmqEventBusConnector
from .RabbitmqQueueNameFormatter      import RabbitmqQueueNameFormatter
from .RabbitmqExchangeNameFormatter   import RabbitmqExchangeNameFormatter

"""
 *
 * Class
 *
"""

class RabbitmqEventBusConfigurer:

    """
     *
     * Attributes 
     *
    """

    __eventBusConnector     : RabbitmqEventBusConnector
    __eventsInformation     : DomainEventsInformation
    __queueNameFormatter    : RabbitmqQueueNameFormatter
    __exchangeNameFormatter : RabbitmqExchangeNameFormatter

    """
     *
     * Methods 
     *
    """

    @inject
    def __init__( 
        self, 
        eventBusConnector     : RabbitmqEventBusConnector,
        eventsInformation     : DomainEventsInformation,
        queueNameFormatter    : RabbitmqQueueNameFormatter,
        exchangeNameFormatter : RabbitmqExchangeNameFormatter
    ) -> object:
        self.__eventBusConnector     = eventBusConnector
        self.__eventsInformation     = eventsInformation
        self.__queueNameFormatter    = queueNameFormatter
        self.__exchangeNameFormatter = exchangeNameFormatter

    def configure( self ) -> None:
        # Variables
        eventName              : str
        queueName              : str
        deadLetterQueueName    : str
        deadLetterExchangeName : str
        # Code
        for eventInformation in self.__eventsInformation.getAll():
            eventName = eventInformation.getEventName()
            self.__declareExchange( eventName )
            if eventInformation.isConsumedEvent():
                queueName = self.__queueNameFormatter.format( eventInformation )
                self.__declareQueue( queueName )
                self.__bindQueue( queueName, eventName )
                deadLetterQueueName = self.__queueNameFormatter.formatDeadLetter( eventInformation )
                deadLetterExchangeName = self.__exchangeNameFormatter.formatDeadLetter( eventName )
                self.__declareQueue( deadLetterQueueName )
                self.__declareExchange( deadLetterExchangeName )
                self.__bindQueue( deadLetterQueueName, deadLetterExchangeName )                        
    
    def __bindQueue( self, queueName : str, exchangeName : str ) -> None:
        # Variables
        channel : Channel
        # Code
        channel = self.__eventBusConnector.getChannel()
        channel.queue.bind(
            queue       = queueName,
            exchange    = exchangeName,
            routing_key = "#",
        )
        channel.close()
    
    def __declareExchange( self, exchangeName : str ) -> None:
        # Variables
        channel : Channel
        # Code
        channel = self.__eventBusConnector.getChannel()
        channel.exchange.declare(
            exchange      = exchangeName,
            exchange_type = 'fanout',
            durable       = True
        )
        channel.close()
    
    def __declareQueue( self, queueName : str ) -> None:
        # Variables
        channel : Channel
        # Code
        channel = self.__eventBusConnector.getChannel()
        channel.queue.declare(
            queue   = queueName,
            durable = True,       
        )
        channel.close()

File: preparation/src/shared/infrastructure/eventBus/RabbitmqEventBusConfigurer.cs
/* 
 * 
 * Libraries
 *
*/

using RabbitMQ.Client;

/* 
 * 
 * Class
 *
*/

namespace preparation.src.shared.infrastructure {

    public class RabbitmqEventBusConfigurer { 

        /* 
         * 
         * Attributes
         *
        */

        private RabbitmqEventBusConnector     eventBusConnector;
        private DomainEventsInformation       eventsInformation;
        private RabbitmqQueueNameFormatter    queueNameFormatter;
        private RabbitmqExchangeNameFormatter exchangeNameFormatter;

        /* 
         * 
         * Methods
         *
        */

        public RabbitmqEventBusConfigurer( 
            RabbitmqEventBusConnector     eventBusConnector,
            DomainEventsInformation       eventsInformation,
            RabbitmqQueueNameFormatter    queueNameFormatter,
            RabbitmqExchangeNameFormatter exchangeNameFormatter 
        ) {
            this.eventBusConnector     = eventBusConnector;
            this.eventsInformation     = eventsInformation;
            this.queueNameFormatter    = queueNameFormatter;
            this.exchangeNameFormatter = exchangeNameFormatter;
        }

        public void configure() {
            // Variables
            string eventName;
            string queueName;
            string deadLetterQueueName;
            string deadLetterExchangeName;
            // Code
            foreach( DomainEventInformation eventInformation in eventsInformation.getAll() ) {
                eventName = eventInformation.getEventName();
                declareExchange( eventName );
                if( eventInformation.isConsumedEvent() ) {
                    queueName = queueNameFormatter.format( eventInformation );
                    declareQueue( queueName );
                    bindQueue( queueName, eventName );
                    deadLetterQueueName    = queueNameFormatter.formatDeadLetter( eventInformation );
                    deadLetterExchangeName = exchangeNameFormatter.formatDeadLetter( eventName );
                    declareQueue( deadLetterQueueName );
                    declareExchange( deadLetterExchangeName );
                    bindQueue( deadLetterQueueName, deadLetterExchangeName ) ;
                }
            }
        }

        private void declareExchange( string exchangeName ) {
            // Variables
            IModel channel;
            // code
            channel = eventBusConnector.getChannel();
            channel.ExchangeDeclare(
                exchange : exchangeName, 
                type     : "fanout", 
                durable  : true
            );
            channel.Close();
        }

        private void bindQueue( string queueName, string exchangeName ) {
            // Variables
            IModel channel;
            // code
            channel = eventBusConnector.getChannel();
            channel.QueueBind(
                queue      : queueName,
                exchange   : exchangeName, 
                routingKey : "#"
            );
            channel.Close();
        }

        private void declareQueue( string queueName ) {
            // Variables
            IModel channel;
            // code
            channel = eventBusConnector.getChannel();
            channel.QueueDeclare(
                queue      : queueName,
                durable    : true,
                exclusive  : false,
                autoDelete : false
            );
            channel.Close();
        }

    }

}